---
title: "Breast Cancer Expression Data"
author: "Diana Murray"
date: "June 15, 2023"
output:
  html_document:
    toc: true
    toc_float: true
    theme: "spacelab"
    code_folding: "show"

---

## About this activity

We will load and examine R data objects that contain data from over 1,000 breast cancer (BRCA) patients from **The Cancer Genome Atlas (TCGA)**. 

The objects include:

* clinical measurements on the patients and the patients' tumors, such as gender, age, estrogen, progesterone, and her2 receptor status. We examined this data in detail in our previous activity, Breast_Cancer_Patient_Data.Rmd.

* gene expression data which tells us how many messenger RNAs (mRNAs) per gene are present in a patient sample. The amount of a gene's mRNA corresponds (roughly) to the amount of protein in the sample.


---

## Loading the data

Like last time, we will load data from R files with the extension `.RData`. `.RData` files are binary files (i.e., not human readable) that can store multiple R objects, such as vectors, lists, matrices, and data frames. 


```{r}

# Create an object where the data is stored
data_dir <- "/home/shared/data"     

# Round floating point numbers
options("digits" = 2)

# We `load()` TCGA_brca.RData
# The file.path() function tells `load()` where our data resides
# The objects will also appear in our "Environment" tab.

load(file.path(data_dir, "TCGA_brca.RData"),verbose=TRUE)  


``` 

---

## Revisiting the clinical data frame

During our last meeting, we spent some time exploring the clinical dataframe, `brca_clin_df`. Let's review its properties.

```{r,eval=FALSE}

View(brca_clin_df)

```


The rows correspond to patients and the columns correspond to clinical features.


```{r}

colnames(brca_clin_df)

```


### Histograms

Another way to look at the values of a feature is a **histogram**. 

```{r}

hist(brca_clin_df$age_at_diagnosis,
     xlab="Age at Diagnosis",
     main="Distribution of Ages")
```

The feature `age_at_diagnosis` follows a normal distribution.  In a **normal distribution**, data are symmetrically distributed with no skew. Most values cluster around a central region, with values tapering off as they go further away from the center. Measures of central tendency (mean and median) are the same in a normal distribution.



```{r}
# remove consideration of missing values
age<-na.omit(brca_clin_df$age_at_diagnosis)

# calculate mean
mean(age)

# calculate median
median(age)

```


---

## Gene expression data 

Another important source of information in The Cancer Genome Atlas (TCGA) breast cancer set is the **gene expression data**.

```{r}

dim_expr <- dim(brca_expr_mat)  # Check the Environment tab, too!

print(paste("The gene expression data matrix has",dim_expr[1],
              "rows and",dim_expr[2],"columns."))

# The function `paste()` allows us to `print()` text and numbers together.

```


Let's see what's in the rows and columns using *indexing* to get 

* rows 1 through 5 (1:5) and 
* columns 1 through 5 (1:5)


```{r}
brca_expr_mat[1:5,1:5] 


```



The matrix contains the expression levels of messenger RNAs (mRNAs) for expressed genes.
Each row corresponds to a gene and each column corresponds to a patient sample. 

The expression and clinical data seem to have information for the same patient samples:

* TCGA-3C-AAAU
* TCGA-3C-AALI
* TCGA-3C-AALJ
* TCGA-3C-AALK


We can check this:

```{r}

# are the two arrays the same?
identical(colnames(brca_expr_mat),brca_clin_df[,1])

```

---

### Average expression


Let's look at the average value of the expression for the first gene (TSPAN6) across the patient samples.


```{r}

# The function `mean()` takes the average of a list of numbers
# We can get the data for TSPAN6 either by column index or its name
mean(brca_expr_mat[1,])
mean(brca_expr_mat["TSPAN6",])

```


<br>

#### High and low expression

What is the average expression of the second gene (TNMD)?

```{r}

# The function `mean()` takes the average of a list of numbers
mean(brca_expr_mat["TNMD",])

```

Average expression is much smaller for TNMD (23.7) than TSPAN6 (996.2).  Does this mean anything?

* The gene TSPAN6 codes for the protein Tetraspanin-6. This protein helps signal events that play a role in cell growth and motility, and its higher level suggests motility of breast cancer cells may be important.

* The gene TNMD codes for the protein Tenomodulin, which is important for the formation of tendons. TNMD is highly expressed in tendons, but lowly expressed in other parts of the body. Its lower level suggests it does not play a role in breast cancer.


It would be tedious to examine each of the 18,351 genes in the expression matrix!!


<br>

#### `apply()` mean to all genes

We can use the function `apply()` to *apply* a function like `mean()` to all rows or columns of our matrix at once.

```{r}

# We "apply" the function mean() to all rows.
# The first argument is the matrix.
# The second argument is 1 for "rows".
# The third argument is the function name, `mean()`

mean.expr <- apply(brca_expr_mat, 1, mean)  
                   
mean.expr[1:10]                            

```


Now that we have the mean expression for all 18K genes, how else may we use this information? 

---



### Plotting average expression

Let's try plotting the mean expression values (we saved these values in the object `mean.expr`) as a histogram.  

```{r}

hist(mean.expr, 
     main="Distribution of Gene Expression Values",
     xlab="Mean Expression")


```

Wow!  Most of the average expression values are relatively small but it's to tell because of a few VERY large values.

How many values in mean.expr are greater than 10,000? 

```{r}

# This is a logical expression
greater_than_10000 <- (mean.expr > 10000)

greater_than_10000[1:10]

```


When we use an expression like "mean.expr > 10000" we get a vector that tells us if the condition is TRUE or FALSE for that gene.  Note that most genes do NOT have expression level greater than 10000.

We can use the **logical condition** to create an objective that has *only* genes with expression level > 10000.

```{r}

# Create an object for expression values > 10,000
mean.expr.high <- mean.expr[mean.expr > 10000]

length(mean.expr.high)

```

Out of 18,351 mean values for mRNA expression, only 201 (1.1%) are greater than 10,000.

Let's zoom in on the average expression values *less than* 10000 with the logical condition "mean.expr < 10000".

```{r}

# Let's look at the "distribution" of mean expression values
# that are less than 10000.

mean.expr.low <- mean.expr[mean.expr < 10000]  

hist(mean.expr.low, 
     breaks=50, 
     main="Distribution of Gene Expression Values",
     xlab="Mean Expression")

```

By zooming in, we see that most genes actually have average expression < 2000.


There are only a few genes that have really large average expression (201 out of 18351 genes). The curve in the histogram has a **long right tail**. 

Nearly all genes have relatively low expression, and very few genes have high expression.  

In other words, our data is **highly skewed**.

Many functions in R assume the data are **normally distributed**, which means their histogram is expected to be bell-shaped. 

As we saw above, the ages of the patients follows a normal distribution.



A **log transformation** can be used to make highly skewed distributions less skewed and sometimes normally distributed. And this can be valuable both for 

* making patterns in the data more interpretable, and 
* helping to meet the assumptions of the statistics underlying our analysis. 


---

## Log transformation


Expression data analysis is usually done with log base 2.  

* Log base 2 of 8 is 3 because 2 * 2 * 2 = 8.

* Log base 2 of 4 is 2 because 2 * 2 = 4.

* Log base 2 of 2 is 1 because 2 = 2.

So the values (8, 4, 2) are *log transformed* to (3, 2, 1).  

Let's consider some larger numbers:
    
```{r}
a <- c(4096, 1024, 512, 64)

b <- log(a,2)

plot(a,b,pch=19,col="red")

```
    
Notice how the scale for `b` (vertical axis) is much more compressed than for `a` (horizontal axis).

There is one caveat.  The log2 of 0 is infinity!  

```{r}

log(0,2)

```



But the log2 of 1 is *well-behaved*.

```{r}
log(1,2)
```



So before taking the log2 of our expression data, we will add 1 to each value to avoid the occurrence of "-Inf" values.  This is standard practice and, as evidenced from the histogram of expression values, it is a small change.


```{r}

# Add 1 to each element, then take the log base 2.
brca_expr_mat.log <- log(brca_expr_mat+1, 2)  

# Look at some values:
brca_expr_mat.log[1:5,1:5]

```
 
 
Let's check out what the curve is for the log-transformed expression values: 

```{r}

mean.expr <- apply(brca_expr_mat.log, 1, mean)  
                                                

hist(mean.expr, breaks=100, main="Distribution of Log_2 Gene Expression Values",
  xlab="Mean Log_2 Expression")

```
 
The log-transformed data is less skewed and even has some bell-shaped character.  

While this isn't a true normal distribution of data (especially at the lower range), it is much closer, and importantly, close enough for some exploratory analyses.

Moving forward, we will be working with the log-transformed data matrix `brca_expr_mat.log`.


---


## Mini-DREAM Challenge 



### Question 1: Breast Cancer Genes


When we explored the clinical data, we found that many tumors are ER+ and HER2+.  What are the mean mRNA expression levels of the associated genes, ESR1 and ERBB2?


```{r,eval=FALSE}

# replace ffff with the function name to get the mean value
# Insert the gene name as index to row to get the value

esr1<-  ffff (brca_expr_mat[ ,])

erbb2<- ffff (brca_expr_mat[ ,])

print(esr1)
print(erbb2)

```


Remember that we found only 201 genes with expression levels above 10,000, so, on average, both ESR1 and ERBB2 are highly expressed.


Let's calculate the median values: 


```{r}
median(brca_expr_mat["ESR1",])
median(brca_expr_mat["ERBB2",])
```


What does it mean when the mean of a distribution is higher than the median?



### Question 2: Skew


Look at the distribution of mRNA expression for either gene across patient samples with a histogram:


```{r,eval=FALSE}

hist(brca_expr_mat[  ,]  )

```


Are the data **positively** or **negatively** skewed?


```{r,eval=FALSE}

# uncomment the assignment to submit your answer

# skew <- "positive"

# skew <- "negative"

```






### Create your R Markdown Report


Let's save our work as an html file!


### The knitr R package 


**knitr()** is the R package that generates reports from R Markdown. R Markdown is the syntax we are using in this document, denoted by the .Rmd file type. With `knitr()` and R Markdown, we can create reports of our work in the form of Word doc, PDF, and HTML files. 

An **R package** bundles together code, data, documentation, and tests, and is easy to download and share with others.


```{r setup, include=FALSE}

library(knitr)
knitr::opts_chunk$set(echo = TRUE)

```

Click the "Knit" button at the top of this window and select "Knit to html."


### Submit your work

```{r}
# Load function to submitting answers to the leaderboard.
scripts_dir <- "/home/shared/R"
source(file.path(scripts_dir, "submission_helpers.R"))

# Log into Synapse
synLoginSecure()

# Submit answers
submission <- submit_module_answers(module = 2)
```


### **Congratulations** 

Now that we've gained some experience working with expression data, we can move onto the next activity and look for interesting patterns.


---

### Practice

**Practice 1**


The function `unique()` will provide the unique values in a list.  

First, how many total values are in the column for estrogen receptor (ER) status? 


```{r}

length(brca_clin_df$estrogen_receptor_status)

```


Then, what are the **unique** values in the estrogen receptor (ER) status column?


```{r}
unique(brca_clin_df$estrogen_receptor_status)
```


Finally, note that we get both the unique values and the number of times they occur with the `table()` function we used last time.


```{r}

table(brca_clin_df$estrogen_receptor_status)

```



**Practice 2**: Find the unique values of other features.


```{r,eval=FALSE}
# type the $ symbol and select a feature (column) to find its unique values
unique(brca_clin_df$prior_dx    )

```



**Practice 3**


Combine code chunks with text to provide more comprehensible results:


```{r,eval=FALSE}

# mean() takes the average of a set of values

avg_TSPAN6_expr <- mean(brca_expr_mat["TSPAN6",])   # We are taking the average of expression values
                                             # for the first row of the matrix

avg_TNMD_expr <- mean(brca_expr_mat["TNMD",])     # We are taking the average of expression values
                                             # for the second row of the matrix


# The function paste() allows us to print text and numbers together.
# The function round() lets us choose how many decimal digits we want to show.


print(paste("Average expression of TSPAN6: ", round(avg_TSPAN6_expr,0)))

print(paste("Average expression of TNMD: ", round(avg_TNMD_expr,0)))


```



**Practice 4**


Sort the genes by their mean expression.



```{r}

# we used the `apply()` function to get mean expression for all genes
mean.expr <- apply(brca_expr_mat, 1, mean)  

# `sort()` sorts the values from low to high
#  use the argument to sort from high to low
sorted_mean.expr<-sort(mean.expr,decreasing=TRUE)

head(sorted_mean.expr)
```



**Practice 5**


Use `apply()` to calculate the variance of the genes across patient samples, i.e. in each row.


```{r}

# `apply()` is used to get the variance ("var") of each row 
# `var()` calculates the variance and is an argument to `apply()`
# The argument "1" denotes calculations on rows

var_genes <- apply(brca_expr_mat.log, 1, var)   

head(var_genes)
                                        
```



