--
title: "Motor-clutch model and Cell Migration"
author: "Jay Hou and Brian Castle"
date: "7/14/2022"
output:
  pdf_document: default
  html_document: default
  html_notebook: default
---
# R code
# indent, select multiple lines, tab

# Preliminaries

```{r setup, include=FALSE, message = FALSE}

library(knitr)
knitr::opts_chunk$set(echo = TRUE)

library(imager)  # We are loading this package so we may look at images.

knitr::opts_chunk$set(echo = TRUE)
data_dir <- "/home/data"
img_dir <- "/home/data/img"
```

## The overall goal of the cell migration activity

We have been learning how to computationally model and simulate a **motor-clutch system** for cell traction dynamics as described by Chan & Odde, Science, 2008.

Last time we examined

1. Mechanisms of cell migration
2. Stochastic modeling of clutch dynamics
3. F-actin binding and stretching clutch

Today, we will delve deeper by considering 

1. Deformable substrates
2. How traction force is related to subtrate stiffness
3. And differential traction between the two breast cancer cell models we considered in Module 2


## clutchs bind on deformable substrate

### governing equation
Now we can apply total clutch forces to a deformable substrate, and the deformation of substrate can affect system dynamics.

We need to track the clutch position
```{r}
knitr::include_graphics(file.path(img_dir,"clutch_x.png"))
```

Based on the force balance, we can derive the substrate position,
```{r}
knitr::include_graphics(file.path(img_dir,"clutch_xsub.png"))
```

Then the net stretch of the clutch is
```{r}
knitr::include_graphics(file.path(img_dir,"clutch_stretch.png"))
```


### Exercise 3: substrate is deformable
For each time step, 

1. stretch bond clutches using constant F-actin flow
2. calculate clutch forces
3. calculate clutch bond and unbound probability
4. bond and unbond the clutches

Now simulate!
```{r}
# parameters
clutchNum <- 75     
kon <- 1       # s^-1
koff <- 0.1    # s^-1
clutchK <- 5       # pN/nm
bondForce <- 2      # pN
subK <- 0.1          # pN/m
actinFlow <- 120

timeStep = 0.01      # seconds
numSteps = 10000;

# Initialize
clutchState <- matrix(0,clutchNum,1)
clutchdx <- matrix(0,clutchNum,1)
clutchx <- matrix(0,clutchNum,1)
clutchF <- matrix(0,clutchNum,1)
moduleF <- 0
subx <- 0
csArray <- matrix(0,numSteps,1)
timeArray <- matrix(0,numSteps,1)
FArray <- matrix(0,numSteps,1)
FlowArray <- matrix(0,numSteps,1)
SubxArray <- matrix(0,numSteps,1)

# The first part of this chunk is the same as before
for (i in 2:numSteps) {

    for (j in 1:clutchNum) {
    
    if (clutchState[j] == 0) {
      pBind = 1 -exp(-kon*timeStep)
      temp <- runif(1)
      if (temp <= pBind) {
        clutchState[j] <- 1 }
    }
    if (clutchState[j] == 1) {
      pUnbind <- 1 -exp(-koff*exp(clutchF[j]/bondForce)*timeStep)
      temp <- runif(1)
      if (temp <= pUnbind) {
        clutchState[j] <- 0 
        clutchdx[j] <- 0;
      }
    }
    if (clutchState[j] == 1) {   # if clutch is bound
      clutchdx[j] <- clutchdx[j] + actinFlow*timeStep   # clutch stretched by flow
    }
    
    clutchx[j] = clutchdx[j] + subx
  }
  
  subx = 0
  for (j in 1:clutchNum) {
    if (clutchState[j] == 1) {
      subx = subx + clutchx[j]
    }
  }
  engNum = sum(clutchState)

  subx = subx*clutchK/(subK+clutchK*engNum)
  
  moduleF = 0
  for (j in 1:clutchNum) {
    if (clutchState[j] == 1) {
      clutchdx[j] = clutchx[j] - subx
    }
    if (clutchState[j] == 0) {
      clutchdx[j] = 0
    }
      clutchF[j] = clutchK*abs(clutchdx[j])
      moduleF = moduleF + clutchF[j]
  }
  
  timeArray[i] = timeArray[i-1] + timeStep
  csArray[i]=engNum     
  FArray[i]=moduleF
  SubxArray[i]=subx
}

plot(timeArray,csArray,type="l",xlim=c(0,max(timeArray)), ylim=c(0,max(csArray)+2),xlab="Time (sec)",ylab="Number of bound clutches",col="blue")

plot(timeArray,FArray,type="l",xlim=c(0,max(timeArray)), ylim=c(0,max(FArray)+2),xlab="Time (sec)",ylab="module force",col="blue")

plot(timeArray,SubxArray,type="l",xlim=c(0,max(timeArray)), ylim=c(0,max(SubxArray)+2),xlab="Time (sec)",ylab="substrate position",col="blue")

ameanF = mean(FArray)

```

This load-and-fail behavior was observed in the experiment.
```{r}
knitr::include_graphics(file.path(img_dir,"load_fail.gif"))
```

, and its measured deformation is similar to simulation.
```{r}
knitr::include_graphics(file.path(img_dir,"exp_load_fail_deformation.png"))
```


**Challenge Questions**
What is the approximate frequency for clutches to unbound (unbound_freq_deform)?

What is the mean module force (mforce_mean_deform)? 

```{r}
# Challenge Question
# unbound_freq_deform <-
# mforce_mean_deform<-
```

**thinking more about the exercise**
What do you observe?

Is substrate deformation = module force * subK? (Equation_SubDeform_MForce = yes or no) 

How will subK affect simulation results?

When subK becomes very large, will the results become similar to Exercise 2?

What is the biological behavior represented by this exercise?



### Exercise 4: the relation between traction force and subtrate stiffness

For each loop, run the simulation with increasing stiffness
For each time step, 
1. stretch bond clutches using constant F-actin flow
2. calculate clutch forces
3. calculate clutch bond and unbound probability
4. bond and unbond the clutches

Now simulate!
```{r}

stiffnessN <- 5
stiffness_arry <- matrix(c(0.01,0.1,1,10,100),stiffnessN,1)
Fmean_array <-matrix(0,stiffnessN,1)

for (k in 1:stiffnessN) {
  # parameters
  clutchNum <- 75     
  kon <- 1       # s^-1
  koff <- 0.1    # s^-1
  clutchK <- 5       # pN/nm
  bondForce <- 2      # pN
  subK <- stiffness_arry[k]          # pN/nm
  actinFlow <- 120
  
  timeStep = 0.01      # seconds
  numSteps = 10000;
  
  # Initialize
  clutchState <- matrix(0,clutchNum,1)
  clutchdx <- matrix(0,clutchNum,1)
  clutchx <- matrix(0,clutchNum,1)
  clutchF <- matrix(0,clutchNum,1)
  moduleF <- 0
  subx <- 0
  csArray <- matrix(0,numSteps,1)
  timeArray <- matrix(0,numSteps,1)
  FArray <- matrix(0,numSteps,1)
  FlowArray <- matrix(0,numSteps,1)
  SubxArray <- matrix(0,numSteps,1)
  
  # The first part of this chunk is the same as before
  for (i in 2:numSteps) {
  
      for (j in 1:clutchNum) {
      
      if (clutchState[j] == 0) {
        pBind = 1 -exp(-kon*timeStep)
        temp <- runif(1)
        if (temp <= pBind) {
          clutchState[j] <- 1 }
      }
      if (clutchState[j] == 1) {
        pUnbind <- 1 -exp(-koff*exp(clutchF[j]/bondForce)*timeStep)
        temp <- runif(1)
        if (temp <= pUnbind) {
          clutchState[j] <- 0 
          clutchdx[j] <- 0;
        }
      }
      if (clutchState[j] == 1) {   # if clutch is bound
        clutchdx[j] <- clutchdx[j] + actinFlow*timeStep   # clutch stretched by flow
      }
      
      clutchx[j] = clutchdx[j] + subx
    }
    
    subx = 0
    for (j in 1:clutchNum) {
      if (clutchState[j] == 1) {
        subx = subx + clutchx[j]
      }
    }
    engNum = sum(clutchState)
  
    subx = subx*clutchK/(subK+clutchK*engNum)
    
    moduleF = 0
    for (j in 1:clutchNum) {
      if (clutchState[j] == 1) {
        clutchdx[j] = clutchx[j] - subx
      }
      if (clutchState[j] == 0) {
        clutchdx[j] = 0
      }
        clutchF[j] = clutchK*abs(clutchdx[j])
        moduleF = moduleF + clutchF[j]
    }
    
    timeArray[i] = timeArray[i-1] + timeStep
    csArray[i]=engNum     
    FArray[i]=moduleF
    SubxArray[i]=subx
  }
  afreq = 8/94
  Fmean_array[k] = mean(FArray)
}

```

```{r}
stiffnesslog = log10(stiffness_arry);
plot(stiffnesslog,Fmean_array,type="l",xlim=c(min(stiffnesslog),max(stiffnesslog)), ylim=c(0,max(Fmean_array)+2),xlab="Log10(stiffness) (pN/nm)",ylab="Force(mean) (pN)",col="blue")
```

This optimal traction has been observed in the experiments.
```{r}
knitr::include_graphics(file.path(img_dir,"CMS_result.png"))
```


**Challenge Questions**
What is the optimal stiffness for the maximum traction force?

```{r}
# Challenge Question
# optimal_stiffness <- 
```

**thinking more about the exercise**
What do you observe?

Why there is the optimal stiffness for the maximum traction force?

What is the biological behavior represented by this exercise?



## Predicting differential traction force between T-47D and MD-MBA-231 cell line

### differential genes between T-47D and MD-MBA-231 cell line
In Module 2, **Breast_Cancer_Cell_Lines**, we derived the differential genes between T-47D and MD-MBA-231 cell line and their mRNA expressions.

We assumed that these differential genes contribute to the physical parameters to predict the differential traction force between T-47D and MD-MBA-231 cell line, which regulate the differential cell migration.

Let's recall the list of genes more highly expressed in MD-MBA-231 cell.

We will start with the matrix we created for the log-transformed gene expression for the two cell lines. 


```{r}

# Load the data we derived from Module 2

load(file.path(data_dir, "hyal_brca_log_mat.RData"),verbose=TRUE)

```


Let's remind ourselves what is there!

```{r}

head(hyal_brca_log_mat)

```

The matrix contains the expression in TPM for each gene in the two cell lines.


```{r}

# Subtract the first and second columns
dge <- hyal_brca_log_mat[,2] - hyal_brca_log_mat[,1]

# Add the column to a new matrix 
DGE_mat <- cbind(hyal_brca_log_mat,dge)

# Sort the differential gene expression values from high to low
order_dge <- order(dge, decreasing = TRUE)

# Re-order the rows of the matrices so we can easily see which
# genes are most expressed in "fast" versus "slow"
DGE_mat <- DGE_mat[order_dge,]
gene_list <- DGE_mat[1:30,1:2]
# Look at the first 10 rows
head(DGE_mat, 30)

```

We apply the enrichment analysis by inserting these gene list into http://geneontology.org, we found VIM is involved in the "focal adhesion" cellular component. T-47D line has the VIM expression of 1.2 and MD-MBA-231 line has the VIM expression of 11. 

We then assume clutchNum is a function of VIM expressions: clutchNum = 100*(VIM expression/6.1), and we derive T-47D line has the clutchNum = 20 and MD-MBA-231 line has the clutchNum = 180.

# Exercise 5: differential traction between T-47D and MD-MBA-231 line

For each loop, run the simulation with increasing stiffness
For each time step, 
1. stretch bond clutches using constant F-actin flow
2. calculate clutch forces
3. calculate clutch bond and unbound probability
4. bond and unbond the clutches

Now simulate!
```{r}

clutchN <- 2
clutch_arry <- matrix(c(20,180),clutchN,1)
Fmean_array <-matrix(0,clutchN,1)

for (k in 1:clutchN) {
  # parameters
  clutchNum <- clutch_arry[k]     
  kon <- 1       # s^-1
  koff <- 0.1    # s^-1
  clutchK <- 5       # pN/nm
  bondForce <- 2      # pN
  subK <- 1        # pN/nm
  actinFlow <- 120
  
  timeStep = 0.01      # seconds
  numSteps = 10000;
  
  # Initialize
  clutchState <- matrix(0,clutchNum,1)
  clutchdx <- matrix(0,clutchNum,1)
  clutchx <- matrix(0,clutchNum,1)
  clutchF <- matrix(0,clutchNum,1)
  moduleF <- 0
  subx <- 0
  csArray <- matrix(0,numSteps,1)
  timeArray <- matrix(0,numSteps,1)
  FArray <- matrix(0,numSteps,1)
  FlowArray <- matrix(0,numSteps,1)
  SubxArray <- matrix(0,numSteps,1)
  
  # The first part of this chunk is the same as before
  for (i in 2:numSteps) {
  
      for (j in 1:clutchNum) {
      
      if (clutchState[j] == 0) {
        pBind = 1 -exp(-kon*timeStep)
        temp <- runif(1)
        if (temp <= pBind) {
          clutchState[j] <- 1 }
      }
      if (clutchState[j] == 1) {
        pUnbind <- 1 -exp(-koff*exp(clutchF[j]/bondForce)*timeStep)
        temp <- runif(1)
        if (temp <= pUnbind) {
          clutchState[j] <- 0 
          clutchdx[j] <- 0;
        }
      }
      if (clutchState[j] == 1) {   # if clutch is bound
        clutchdx[j] <- clutchdx[j] + actinFlow*timeStep   # clutch stretched by flow
      }
      
      clutchx[j] = clutchdx[j] + subx
    }
    
    subx = 0
    for (j in 1:clutchNum) {
      if (clutchState[j] == 1) {
        subx = subx + clutchx[j]
      }
    }
    engNum = sum(clutchState)
  
    subx = subx*clutchK/(subK+clutchK*engNum)
    
    moduleF = 0
    for (j in 1:clutchNum) {
      if (clutchState[j] == 1) {
        clutchdx[j] = clutchx[j] - subx
      }
      if (clutchState[j] == 0) {
        clutchdx[j] = 0
      }
        clutchF[j] = clutchK*abs(clutchdx[j])
        moduleF = moduleF + clutchF[j]
    }
    
    timeArray[i] = timeArray[i-1] + timeStep
    csArray[i]=engNum     
    FArray[i]=moduleF
    SubxArray[i]=subx
  }
  afreq = 8/94
  Fmean_array[k] = mean(FArray)
}

```

```{r}
cellarr <- c("T-47D","MD-MBA-231")
barplot(Fmean_array[1:2,1],names.arg=cellarr, ylab="traction force (pN)")
```

The vimentin-knockdown cell line showed lower traction force than control 3T3 fibroblasts.

```{r}
knitr::include_graphics(file.path(img_dir,"VIM.png"))
```

**Challenge Questions**

What is the ratio between the traction force of MD-MBA-231 and T-47D cell line?

```{r}
# Challenge Question
# traction_ratio <-
```

**thinking more about the exercise**
What do you observe?

What is the biological behavior represented by this exercise?



### stochastic modeling of motor-clutch mechanism
```{r}
knitr::include_graphics(file.path(img_dir,"motor_clutch_model.png"))
```

Stochastic simulations of the motor-clutch model (Chan & Odde, Science, 2008) has been developed and applied to predict the cell traction force, cell morphology, and F-actin flow at different substrate conditions. 
  Module force Fj is summation of individual clutch force Fc,i, which is a function of the clutch spring constant kc and clutch displacements dxc,i. Clutches bound and unbound to F-actin based on the clutch binding rate kon  and unbinding rate koff  as a function of the minimum unbinding rate koff, motor force Fm  and Fj, and Fm  equals to the single motor force Fm multiplying the motor number nm. Bound clutches were displaced at the rate vactin as a function of the maximum rate vactin*, nm , Fm , and F_j.
  Monte Carlo simulations were conducted using a direct Gillespie Stochastic Simulation Algorithm, the event was executed based on accumulated event rates, including kon, koff, and the next time step t_step was determined based on the total event rates and a random number RN.
  the simulation results were consistent with experimental results for neurons contracting on polyacrylamide gel.
  
```{r}
knitr::include_graphics(file.path(img_dir,"motor_clutch_results.png"))
```



### stochastic modeling of cell migration
```{r}
knitr::include_graphics(file.path(img_dir,"CMS.png"))
```

By imposing a force balance between the protrusions, Bangasser, Nat Commun, 2017 successfully developed a physics-based cell migration simulator (CMS) to simulate the optimal substrate stiffness for the maximum motility of U251 glioma cells on PA gels, and subsequent work predicted the optimal cell motility with different focal adhesion sizes and distributions (Hou, Cytoskeleton, 2019), simulated the higher cancer cell motility on viscoelastic gels with faster stress relaxing (Adebowale, Gong, Hou, Nat Mat, 2021), predicted the negative durotaxis for U251 glioma cells on PA gels with stiffness gradients (Isomursu', Park', Hou', Nat Mat, 2022), and simulated the cyclic cell migration speed within 1D channels for melanoma cells (Lee, Hou, Cell Systems, 2022). By changing parameter values, the CMS can capture the differential migration features in different cell types and under drug treatments, which may link to cell transcriptomes.  Therefore, the CMS provides a consistent mechanical framework which can be used to interpret and synthesize cell migration and force measurements across glioblastoma subtypes in order to predict cell migration based on glioblastoma cell transcriptomes.

The CMS comprises multiple protrusions or modules which were nucleated randomly based on the rate kmod  which is a function of the maximum rate km', G-actin or free actin monomers, and the total actin amount (AT). Protrusions were elongated based on the polymerization rate vpoly as a function of the maximum rate vpoly' , AG, and AT. Protrusions were capped randomly at the rate kcap  eliminating further polymerization and removed if the protrusion length is shorter than the minimum length lmin. The cell position was determined by the force balance between protrusion forces Fj for nmod  modules and the cell body force Fcell.  Fcell  and Fj  are summation of individual clutch force Fc,i, which is a function of the clutch spring constant κc  and clutch displacements dxc. The rest of equations follow the motor-clutch model.
	Monte Carlo simulations were conducted using a direct Gillespie Stochastic Simulation Algorithm, the event was executed based on accumulated event rates, including kon, koff, kmod, and kcap, and the next time step tstep  was determined based on the total event rates and a random number RN. The C++ version of the CMS (Hou, Cytoskeleton, 2019) was used to conduct the simulations in the Mesabi computer cluster at the Minnesota Supercomputing Institute (MSI).

  the simulation results were consistent with experimental results for U251 glioma cells on the polyacrylamide gel
```{r}
knitr::include_graphics(file.path(img_dir,"CMS_result.png"))
```


### Submitting the prediction

You're now ready to submit the prediction. Just run the chunk below, a file with your prediction will be uploaded to Synapse and submitted to the challenge. You'll be able to see the results of your prediction on the mini-DREAM scoreboards, with the submission ID that gets printed below.


Submit your answers to Synapse
```{r}
# Load function to submitting answers to the leader board
scripts_dir <- "/home/shared/R"
source(file.path(scripts_dir, "submission_helpers.R"))

# Log into Synapse
synLoginSecure()  # You might be prompted for your username and password

# Submit answers
submission <- submit_module_answers(module = 6)
```



### the end

Congrats — you’ve reached the end of the module!

**think more about the module**
Does this module help you to understand better about cell migration?



