---
title: "Module 5: Integrating Biological Knowledge"
author: "Kamrun Begum and Diana Murray"
date: "July 12, 2021"
output: html_notebook
editor_options: 
  chunk_output_type: inline
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

data_dir <- "/home/rstudio/shared/data"

scripts_dir <- "/home/rstudio/shared/R"
source(file.path(scripts_dir, "submission_helpers.R"))
```

# About this activity

## Enrichment analysis
We will use the R package **clusterProfiler** to perform enrichment analysis of GO annotations for sets of genes we think may be related to motility. We started this analysis with Andrew in **Module 2**. 

We know from experience that we most likely do not need to write our own code to perform the analyses. But how do we know where to find trustworthy code?

## Bioconductor for R packages like `clusterProfiler`
The clusterProfiler package is maintained through [**Bioconductor**](https://bioconductor.org/), an open-source/open-development platform for bioinformatics tools in R. A large community of professionals supports the development, maintenance, and use of Bioconductor. Most of the packages available on Bioconductor are available on the github platform as well, but Bioconductor provides more comprehensive community support.

There are a number of ways to [install Bioconductor packages into your local RStudio environment](https://bioconductor.org/install/).

At the top of the [Bioconductor page for clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html), there are some metrics that describe how widely used and trustworthy the package is — e.g., its download profile (rank = 36/2041, in top 2% of downloaded packages), the support activity (most posted questions received answers), and a multi-platform check report ("all"). clusterProfiler is also supported by a couple of publications from [20212](https://www.ncbi.nlm.nih.gov/pubmed/22455463) and [2021](https://pubmed.ncbi.nlm.nih.gov/22455463/). It is also [available on github](https://guangchuangyu.github.io/software/clusterProfiler/). Overall, clusterProfiler looks like a solid package.

Available are a manual with extensive documentation and **Vignettes**, which include examples of implementing clusterProfiler. The requirement of Vignettes is an outstanding feature of Bioconductor packages.  We will use only a few of the functions included in clusterProfiler. If you want to learn more about the package, you can work through [the clusterProfiler Vignettes](https://bioconductor.org/packages/devel/bioc/vignettes/clusterProfiler/inst/doc/clusterProfiler.html). 

Another package we could use is **gprofiler2**.  You can learn how to use this package from a [**great Nature Protocols paper**](https://pubmed.ncbi.nlm.nih.gov/30664679/).


## Bioconductor packages for this activity
1) [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html) for functional enrichment (or overrepresentation) analysis
2) [org.Hs.eg.db](https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html) for help with gene annotation information https://bioconductor.org/packages/release/data/annotation/html/org.Hs.eg.db.html
3) [DOSE for Disease Ontology Semantics and Enrichment Analysis](https://bioconductor.org/packages/release/bioc/html/DOSE.html)


## Loading the packages for this activity
When we install packages on our local machines, we have to call them with the **`library()`** function.

```{r, message=FALSE, warning=FALSE}
library(clusterProfiler)   
library(org.Hs.eg.db)      
library(DOSE)
```


## Loading & formatting data
We revisit the work we did in **Module 2** for loading and manipulating the PS-ON cell line expression and motility data.

Run the following chunk to load the data and to match (`merge()`) with gene information.

```{r}
load(file.path(data_dir, "pson_expr_tpm_df.RData"))
load(file.path(data_dir, "pson_expr_gene_info.RData"))
load(file.path(data_dir, "pson_motility_tidy_df.RData"))

# We add gene information by merging the expression and id data frames
pson_expr_tpm_df2 <- merge(           
  x = gene_df, y = pson_expr_tpm_df,
  by.x = "gene_id", by.y = "gene_id"
)

pson_expr_tpm_df2[1:5,1:10]           # We use `indexing` to inspect 
```

We saw in earlier modules that there are many ways to identify genes. For example, human KRAS has the Ensembl id: "ENSG00000133703"; the hugo gene ID: KRAS; the UniProt ID: P01116; and the Entez ID: 3845. 

We will use the Ensembl `gene_id` (column 1) in our enrichment analysis, and we'll keep the gene name `symbol` (column 3) so we can easily identify genes with their common names.

The following chunk creates the expression matrix for 18K+ genes across the ~60 cell lines.

```{r}
gene_symbol<-pson_expr_tpm_df2$symbol  # We'll reserve the gene symbols 

# Remove the first 5 columns from the merged data frame and convert to a matrix
pson_tpm_mat <- as.matrix(pson_expr_tpm_df2[, -c(1:5)])  

# We rename the matrix to remind ourselves with gene id we're using for analysis
pson_tpm_mat_Ensembl <- pson_tpm_mat
rownames(pson_tpm_mat_Ensembl) <- pson_expr_tpm_df2$gene_id

# Log-transformed expression matrix
pson_logtpm_mat_Ensembl <- log2(1 + pson_tpm_mat_Ensembl)

pson_logtpm_mat_Ensembl[1:5, 1:5]   # The expression matix has 
                                    # Ensembl identifiers as row names
                                    # and cell line codes as column names
```


## Hyaluronic acid

We are interested in cell motility, so we are considering aspects of cell measurements, like cell speed, in addition to gene expression. Here is a cartoon depiction of a cell in its biological environment. Extracellular matrix proteins include collagen (light blue), hyaluronic acid (dark blue), and fibronectin (green). These proteins are used in vitro as substrates on which to measure cell properties. In this activity, we'll look at cell speed measurements on the **hyaluronic acid substrate**.

```{r}
library(imager)

substrate_img<-load.image(file.path(data_dir,"substrate.JPG"))

plot(substrate_img,axes=FALSE)
```

Before we collect cell speed data, let's remind ourselves what the PSON physical data includes.

```{r}
View(pson_motil_tidy_df)
```

Cell speed is in um/hr, and there are measurements for different human cell lines on different substrates.

This code chunk extracts the cell speed data for the two breast cancer cell lines on the hyaluronic acid collagen substrate. We did this with Andrew in **Module 2**. 

```{r}
cell_speeds_df <- subset(pson_motil_tidy_df, summary_metric == "speed_um_hr")

hyal_coll_df <- subset(
  cell_speeds_df,
  experimentalCondition == "HyaluronicAcid Collagen"
)

# We want to work with the samples that also have expression data
samples_w_expr <- match(hyal_coll_df$sample, colnames(pson_logtpm_mat_Ensembl))

hyal_coll_logtpm_mat_Ensembl <- pson_logtpm_mat_Ensembl[, samples_w_expr]

hyal_coll_df$cellLine[which(hyal_coll_df$diagnosis == "Breast Cancer")]

breast_cancer_cols<-which(hyal_coll_df$diagnosis == "Breast Cancer")
```

Last time, we considered T-47D as a model for less aggressive breast cancer and MDA-MB-231 as a model for more aggressive breast cancer.

---

## Associating genes with differences in motility

In this chunk, we create a data frame that contains the ranked difference in expression between the two breast cancer cell lines, **T-47D** and **MDA-MB-231**.  Very different average speeds are observed on the hyaluronic acid collagen substrate: 16.5 u/hr for T-47D and 36.2 u/hr for MDA-MB-231.

```{r}
breast_logtpm_df <- data.frame(
  rownames(hyal_coll_logtpm_mat_Ensembl),  # Ensembl gene ids are required for our analysis
  gene_symbol,     # We include gene symbols since these are meaningful to us
  hyal_coll_logtpm_mat_Ensembl[, breast_cancer_cols]
)

names(breast_logtpm_df) <- c("Ensembl", "symbol", "slower", "faster")

breast_logtpm_df$delta <- breast_logtpm_df[, "faster"] -
  breast_logtpm_df[, "slower"]
breast_logtpm_df <- breast_logtpm_df[order(breast_logtpm_df$delta, decreasing = TRUE), ]

rownames(breast_logtpm_df) <- NULL

head(breast_logtpm_df, 10)
```

The genes at the top of this list show much greater expression in the *faster* cell line (MDA-MB-231); if we had ordered the list by `decreasing = FALSE`, then we would have negative delta values at the top and be looking at genes with greater expression in the *slower* cell line (T-47D).  (You can try this!)

Since we are interested only in `delta` (expression in the faster cell line - expression in the slower cell line), we'll remove the columns for 'slower' and 'faster`.

```{r}
breast_logtpm_df2 <- breast_logtpm_df[, -c(3,4)]

```


Now we'll set up the gene list we need for performing the GO enrichment analysis.

`geneList_ensembl` is a ranked list of "delta" values (the difference in expression for each gene in the "fast" versus "slow" condition) with the Ensembl gene IDs as names. `clusterProfiler` likes to work with Ensembl IDs because they are stable, unique, and unambiguous.  `geneList_symbol` is the same ranked list but annotated with gene symbols for our reference.

```{r, warning=FALSE}
geneList_ensembl <- breast_logtpm_df2[, 3]
names(geneList_ensembl) <- breast_logtpm_df2[, 1]

geneList_symbol <- breast_logtpm_df2[, 3]
names(geneList_symbol) <- breast_logtpm_df2[, 2]

# Top genes faster in MDA-MB-231 than in T-47D 
head(geneList_symbol,20)
# Top genes faster in T-47D than in MDA-MB-231 
tail(geneList_symbol,20)
```

## Vimentin and keratin in motility

With Andrew in **Module 2**, we inspected the genes at the top and bottom of this ranked list. Andrew noted that some of the genes most differentially expressed were lotto winners in terms of potential association with motility!

For example VIM, at the very top of the list codes for the protein vimentin, and KRT23, near the bottom codes for the protein keratin. 

This image comes from the paper, ["Vimentin induces changes in cell shape, motility, and adhesion during the epithelial to mesenchymal transition"](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2874471). 

Cells that express vimentin intermediate filaments (VIF) but not keratin intermediate filaments (KIF) are elongated in shape and more motile (panel A), whereas cells with KIF and not VIF are round and undergo fewer changes in morphology and position (panel B).  This indicates that vimentin plays a role in cell migration and more aggressive cancer.

```{r}
VIF_KIF_img<-load.image(file.path(data_dir,"VIF_KIF.JPG"))

plot(VIF_KIF_img,axes=FALSE)
```

## Choosing our genes for enrichment analysis

To focus on the genes with the strongest apparent relationship to motility, we'll select only those at the top or bottom of the list. Choosing a cutoff in these situations can often be a bit arbitrary... We can use the `hist()` function to check the distribution of `delta` values and try to pick a threshold that includes a reasonable fraction of the "extreme" values.

```{r}
hist(geneList_ensembl, xlab = "Differential gene expression, delta")
abline(v = c(-4, 4), lty = 2)
```

Again, somewhat arbitrary, but we'll go ahead and use 4 as our cutoff for `delta` in both directions.

We create two sets of genes: 1) Those with "delta" less than or equal to -4 (genes with higher expression in the slow cell line (T-47D)), and 2) those with "delta" greater than or equal to +4 (genes with higher expression in the fast cell line (MDA-MB-231)).  

Remember that the expression values are log2-transformed.  So genes with `delta` = 4 differentially expressed by 16-fold.


```{r}
gene_ensembl_neg <- names(geneList_ensembl[(geneList_ensembl <= -4)])
gene_ensembl_pos <- names(geneList_ensembl[(geneList_ensembl >= 4)])

gene_symbol_neg <- names(geneList_symbol[(geneList_symbol <= -4)])
gene_symbol_pos <- names(geneList_symbol[(geneList_symbol >= 4)])

length(gene_ensembl_pos)
length(gene_ensembl_neg)


```

These are relatively short lists to work with.  


## Gene Ontology (GO) enrichment

The **clusterProfiler** package provides functions for performing a wide range of enrichment analyses.  We will focus on **`enrichGO`**, which annotates the genes in our list with GO (Gene Ontology) terms and then looks for GO annotations that are significantly over- or under-represented.

```{r}
args(enrichGO)
```

Look at the arguments to the function. Note that the `pAdjustMethod` is set to `"BH"`. The Benjamini-Hochberg (BH) procedure decreases the false discovery rate in analysis. Adjusting the rate helps to control for the fact that sometimes small p-values (less than 5%) happen by chance, which could lead us to incorrectly reject the true null hypotheses. In other words, the BH Procedure helps us avoid Type I errors (false positives).

`ont` is short for **ontology** and can be one of "BP", "MF", or "CC".

What should we use as our reference gene set or "universe"?  

Check out the documentation for the `enrichGO()` function to learn a bit more about how it works.

```{r}
?enrichGO
```

We'll first look for enriched **GO biological process terms** with the argument `ont = "BP"`. 
Running the function will take a few moments.

Examine the top entries of our results by scrolling horizontally through the columns and connect the output with the information we obtained Tuesday with the [online GO Panther enrichment tool](http://geneontology.org/). 

```{r}
ego_pson_pos <- enrichGO(
  gene = gene_ensembl_pos,
  universe = names(geneList_ensembl),
  OrgDb = org.Hs.eg.db,
  keyType = "ENSEMBL",
  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_pos,12)
```

Are these annotations consistent with our "hypothesis" that genes with the greatest differential expression between the "fast" and "slow" breast cancer cell lines may be related to motility?

We can use the visualization tools **clusterProfile** provides to more easily inspect our results.  Run the following chunk.  Do you prefer the output table above or the plot?


```{r}
dotplot(ego_pson_pos,font.size=10,showCategory=8)
```


Let's look at the enrichment of annotations for genes at the bottom of the rank list — i.e. those genes that are under-expressed in "fast" versus "slow" breast cancer cell lines.

```{r, warning=FALSE}
ego_pson_neg <- enrichGO(
  gene = gene_ensembl_neg,
  universe = names(geneList_ensembl),
  OrgDb = org.Hs.eg.db,
  keyType = "ENSEMBL",
  ont = "CC",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_neg)
```

So the genes at the bottom of our list- the genes overexpressed in the slower moving cells- clearly like to participate in tissue building.


```{r}
dotplot(ego_pson_neg,font.size=10)

```

## Practice

Recall that GO has three main aspects:

1. biological process (BP),
2. molecular function (MF),
3. cellular component (CC). 

We already looked at biological process (BP) annotations in the previous two examples.  Use the `enrichGO()` function, provided again in the next chunk, to examine the enrichment of "MF" and "CC" annotations.  

```{r, warning=FALSE}
ego_pson_pos <- enrichGO(
  gene = gene_ensembl_pos,
  universe = names(geneList_ensembl),
  OrgDb = org.Hs.eg.db,
  keyType = "ENSEMBL",

  # To change the GO aspect, change the ont argument.  
  # It can be "BP", "MF", or "CC"

  ont = "BP",
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_pos)
```

Use barplot or dotplot to visualize the enrichments.

```{r}
# visualize your results with dotplot() function

```


## Gene Set Enrichment Analysis (GSEA)

### Enrichment analysis review
Our GO enrichment or over-representation analysis investigated the fraction of our gene list that appears in a specific GO definition or "GO term." If, say, 0.3% of all human genes are in a pathway, but 20% of the genes in our list are within this pathway, then that GO term is over-represented.  We considered a small subset of genes (1%, or 200 of 18.6K total) with the highest absolute `delta` (> 4). However, this choice is arbitrary. Further, many genes contribute to the dysregulation of cellular pathways in cancer and not all of them are as highly differentially expressed.

### Gene set enrichment analysis
GSEA (gene set enrichment analysis) works differently from over-representation analysis. For GSEA, we provide ALL the genes in our analysis, i.e. without a cutoff. It's important that the genes are ranked by some statistic, for example, differential expression between two cell lines, as we've already done.  In essence, the algorithm walks through the list from top to bottom, investigating a certain pathway (represented by a gene set). When a gene is in the pathway the enrichment score is increased, if it isn't the enrichment score is decreased. The increase depends on the rank and weight, such that the higher the rank and fold change, the bigger the boost in enrichment score. The end result is a single enrichment score that is high if the pathway genes are clustered at the top/bottom, and low if they are randomly distributed throughout the ranked list.

Using the ranked list of differentially expressed genes (geneList_ensembl) we will run Gene Set Enrichment Analysis. The package clusterProfiler provides a function `gseGO()` for this task. The reference gene sets available in this function are the Gene Ontology categories: Biological Process (BP), Cellular Compartment (CC), and Molecular Function (MF).

In the code below we will be using the Biological Process (BP) reference gene set. 

```{r, warning=FALSE}

set.seed(38737)
gsea_BP_BH <- gseGO(geneList=geneList_ensembl, 
                 ont ="BP", 
                 keyType = "ENSEMBL", 
                 minGSSize = 15,
                 maxGSSize = 500,
                 pvalueCutoff = 0.05, 
                 verbose = FALSE, 
                 OrgDb = org.Hs.eg.db, 
                 pAdjustMethod ="BH",
                 seed = TRUE,
                 )
```


Observe the first few enriched gene sets in the GSEA results. How does the list compare with our results from GO Enrichment Analysis?  Scroll horizontally to see the statistics provided by the analysis.

```{r}
head(gsea_BP_BH)
```

To visualize the enriched gene sets, we use the function named `dotplot()` and combine it with another function `facet_grid()`, which allows simultaneous visualization of enriched gene sets of the fast (MDA-MB-231) and slow (T-47D) cell lines. In this plot the fast cell line is represented in the plot labeled "activated" and the slow cell line is represented in the plot labeled "suppressed."  

For each plot the y-axis displays gene set names and the x-axis displays the gene ratio. Gene ratio is the fraction of differentially expressed genes in a gene set over the total number of differentially expressed genes. 

```{r, message= FALSE, warning= FALSE}
dotplot(gsea_BP_BH, showCategory=10, split=".sign", font.size=10) + facet_grid(.~.sign)
```

We obtain results consistent with those we obtained from the less comprehensive over-representation analysis, suggesting our annotations are biologically meaningful.  For example, "activated" is annotated with "positive regulation of cell migration," and "repressed" is annotated with "keratinocyte differentiation" and "epidermal cell differentiation." 

We can summarize the GSEA results of the fast cell line into three categories.
1. modes of movement
  * epithelium migration
  * ameoboid-type cell migration
2. processes involved in movement
  * extracellular matrix organization
  * extracellular structure organization
3. processes that rely on movement
  * angiogenesis
  * gastrulation


Select another sub GO ontology (MF or CC) and run GSEA:

```{r, warning= FALSE, message= FALSE}

my_go_ont = " "

set.seed(121)
my_gsea <- gseGO(geneList=geneList_ensembl, 
                ont = my_go_ont, 
                keyType = "ENSEMBL", 
                minGSSize = 15, 
                maxGSSize = 500, 
                pvalueCutoff = 0.05, 
                verbose = FALSE, 
                OrgDb = org.Hs.eg.db, 
                pAdjustMethod = "none",
                seed = TRUE)

```


Which genesets are over-represented in your GSEA run?

```{r}
head(my_gsea)
dotplot(my_gsea, showCategory=10, split=".sign", font.size=10) + facet_grid(.~.sign)
```


In addition to reference genesets from Gene Ontology, the clusterProfiler package also allows users use other gene sets, for example from [Molecular Signatures Database](https://www.gsea-msigdb.org/gsea/msigdb/). The MSigDB collection of gene sets were designed to reflect pathways and processes involved in cancer. 

```{r, warning= FALSE, message= FALSE}

hallmarks <- read.gmt(file.path(data_dir,"h.all.v7.4.symbols.gmt"))

head(unique(hallmarks$term))

```



```{r, warning= FALSE, message= FALSE}
set.seed(10)
gsea_hall <- GSEA(geneList_symbol, 
                  TERM2GENE=hallmarks, 
                  verbose=FALSE,
                  minGSSize = 15, 
                  maxGSSize = 200, 
                  pvalueCutoff = 0.05, 
                  seed = TRUE)

head(gsea_hall)

```


Plot GSEA results using Hallmark genes. 

```{r, warning= FALSE, message= FALSE}
dotplot(gsea_hall, showCategory=15, split=".sign", font.size=10) + facet_grid(.~.sign)
```

One of the top processes for "activated" ranking "hallmark_epithelial_mesenchymal_transtion". 

"Repressed" ranking shows estrogen responses as top processes, consistent with the T-47D being a model for less aggressive breast cancer.


---

## mini-DREAM Challenge

**QUESTION:** How does enrichment analysis look when we use the ranked absolute values of delta? Consider BP, MF, or CC.

In this activity, you will provide answers in six R objects:
* my_gene_count
* my_go_subontology
* my_top_go_id
* my_go_description
* my_fav_go_term
* my_rationale

First, let's set up the data to use as input for the `enrichGO` function.

```{r}
breast_logtpm_df3 <- breast_logtpm_df2   #  We will rename and use the data frame we
                                         #  created in the code chunks at lines 134 and 156

head(breast_logtpm_df3)
tail(breast_logtpm_df3)

```

So we have positive and negative values for delta.

```{r}
# Replace `aaa` with a function that will return the absolute values of delta.
# HINT: 'a' is the first letter of the function name...
breast_logtpm_df3[, 3] <- aaa(breast_logtpm_df3[, 3])

# Replace ?? with T or F.  How should the absolute values of delta be ranked? 
breast_logtpm_df3 <- breast_logtpm_df3[order(breast_logtpm_df3$delta, decreasing = ??), ]
head(breast_logtpm_df3)

geneList_abs <- breast_logtpm_df3[, 3]
names(geneList_abs) <- breast_logtpm_df3[, 1]

# Replace `nnnnn` with the function that will provide 
# the gene list information you need.
# HINT: 'n' is the first letter of the function name...
gene_pson_abs <- nnnnn(geneList_abs[(geneList_abs >= 4)])
```


Check the length of the resulting gene list, `gene_pson_abs` either using a command you've seen previously or by looking in the environment panel.

```{r}
# Uncomment the following line when you choose your function
# ffffff(gene_pson_abs)

# Replace NNN with the number of genes in the list
my_gene_count <- NNN
```

Now, pick a GO aspect — or "subontology" — and perform the enrichment analysis. Inspect the results and answer a few questions below.

```{r}
# set the GO aspect you want to use with `my_subontology`:
# replace NN with "BP", "MF", or "CC"
my_go_subontology <- "NN"

# Use one of the enrichGO implementations above to perform the analysis
ego_pson_abs <- enrichGO(
  gene = gene_pson_abs,
  universe = names(geneList_abs),
  OrgDb = org.Hs.eg.db,
  keyType = "ENSEMBL",
  ont = my_go_subontology,
  pAdjustMethod = "BH",
  pvalueCutoff = 0.01,
  qvalueCutoff = 0.05,
  readable = TRUE
)

head(ego_pson_abs)
```

What did you find to be the most significantly enriched GO term in your list of genes, based on adjusted p-value?

```{r}
# fill in the GO term ID (e.g., "GO:0012345")
my_top_go_id <- " "

# fill in the description of the GO term 
# (you can copy/paste this from the table)
my_go_description <- " "
```


Use the visualization tool from above to interpret your results.

```{r}
# replace fff in the function name to create your plot as we did in line 265
fffplot(ego_pson_abs)
```

After looking at the plot, do you think that the top GO term from the table above is the most interesting? Which terms, if any, might be more informative of the biological mechanisms that drive motility differences between the two cell lines? Why?

```{r}
# fill in the name of your "favorite" GO term from the plot (use the values
# on the y-axis) — this could be the same or different from the 'go_description'
# you reported above
my_fav_go_term <- " "

# fill in your reason for choosing this GO term
my_rationale <- " "
```


## Submitting the prediction

You're now ready to submit the prediction. Just run the chunk below, a file with your prediction will be uploaded to Synapse and submitted to the challenge. You'll be able to see the results of your prediction on the mini-DREAM scoreboards, with the submission ID that gets printed below.

```{r}
synLoginSecure()  # You might be prompted for your username and password
submission <- submit_module_answers(module = 5)
```

Congrats — you've reached the end of **Module 5**! You can now return to the **mini-DREAM Challenge** site on Synapse.

---

# Bonus — Thought questions

How many genes are in the sets we analyzed?  What would be the consequences of including more or fewer genes in our sets?  What are the potential perils of having gene sets that are "too small" or "too large"?  Would you expect to obtain many statistical enrichments for a small set?  What might result when the set is too large?



