---
title: "mini-DREAM Breast Cancer Patient Data"
author: "Diana Murray"
date: "June 4, 2023"
output:
  html_document:
    toc: true
    toc_float: true
    theme: "spacelab"
    code_folding: "show"
---

In this activity, we will learn new skills in R with a large real-life dataset!

We will load and examine a data frame that contains clinical information from over 1,000 breast cancer patients from *The Cancer Genome Atlas (TCGA)*. 


TCGA characterized over 20,000 cancer samples spanning 33 cancer types with genomics. 

Throughout this course, we will examine some of the different data types and the computational analyses that were performed to decipher this data.  

---

## Preliminaries

These are the tasks we need to do before we begin our analysis.


### Load packages

Many users and developers contribute **packages** to R that contain useful functionality not available with the default R distribution.  At the beginning of our activities, we load the special packages we need.

**Click the green arrow in the top right corner of the *code chunk* to run the code**

```{r,include=FALSE}
# imager provides functions for reading and displaying images
# Annotating our documents with pictures enhances communication
library(imager)

```


### The data directory

We will create an object that holds the name of the directory where the TCGA data resides. This is good programming practice: We can examine data in a different location by changing the directory once at the beginning of our code.


```{r}
# The data we are using is located in /home/data on our cloud server.
data_dir <- "/home/data"     

# Look in your Environment tab!
```

---

## Breast Cancer Basics


Breast cancer most often begins in the ducts.


```{r,warnings=FALSE}
# Note that objects in R can be images.
# Images are matrices of pixel values.

# The `file.path()` function tells `load.image()` where our image resides.
anatomy_img<-load.image(file.path(data_dir,"breast_anatomy.jpg"))

# `par()` defines parameters for plotting
# The argumant `mar` accepts four values
par(mar=c(2,2,2,2))

plot(anatomy_img,axes=FALSE)

```


In breast cancer, the cells lining the ducts grow and divide abnormally.


```{r,warnings=FALSE}
# The `file.path()` function tells `load.image()` where our image resides.
duct_cells_img<-load.image(file.path(data_dir,"duct_cells.jpg"))

par(font.sub=4, mar=c(2,2,2,2))

plot(duct_cells_img,axes=FALSE)

```

We will examine patient data for these ductal breast cancer types.


Metastasis involves the development of secondary malignant growths at a distance from a primary site of cancer.

```{r,warnings=FALSE}
# The `file.path()` function tells `load.image()` where our image resides.
meta_img<-load.image(file.path(data_dir,"metastasis.jpg"))

par(font.sub=4, mar=c(2,2,2,2))

plot(meta_img,axes=FALSE)

```


---

## Loading data

Throughout the course, we will be loading data from R files with the extension `.RData`. 
An example is the file `brca_clin.RData` which we will work with in this activity.

`.RData` files are binary files (i.e., not human readable) that hold R objects, such as data frames. 


```{r}
# With the function `load()` we read in the data file "brca_clin.RData".
# The `file.path()` function tells `load()` where our data resides.
# The argument "verbose = TRUE" makes `load()` tell us what R objects are being loaded.
# The objects will also appear in our "Environment" tab.

load(file.path(data_dir, "brca_clin.RData"),verbose=TRUE)   

```

The object `brca_clin_df` was named to be be descriptive.

* "brca" stands for "TCGA breast cancer"
* "clin" is short for "clinical data"
* "df" stands for "data frame"

Data frames are data displayed as a table. 

In data science, columns in a data frame are features and rows are records or observations.

---

## Viewing the data

The **Environment** tab gives us basic information on the objects that were loaded.

There are 1082 observations (patients) with 27 variables (clinical features). 

We can use some simple R functions to examine the data in more detail.

* `dim()` tells us the dimensions (# rows, # columns) of the object.
* `head()` returns the first several rows of the objects.
* indexing allows us to look at rows and columns we choose.


```{r}
# How many rows and columns are in our object?
dim(brca_clin_df)

```

Is the output what we expect? Check your Environment tab. 

```{r,eval=FALSE}
# Let's take a first look at the data in our object
# `head` gives us the first six rows and ALL of the columns
head(brca_clin_df)

```


Scroll through the columns by clicking on the arrow in the top right corner of the table.


We can see the entire data frame more conveniently in a new tab: 

```{r,eval=FALSE}
View(brca_clin_df)

```


Notice the column names. Let's define a few:

 + `bcr_patient_barcode`  provides anonymous identifiers for the patients.
 + `histological_type` describes the ductal invasion.
 + `estrogen_receptor_status`  designates whether the tumor is ER+ or ER-.
 

Use indexing to view specific elements of the data frame:

```{r}
# Look at a single element
# The first number denotes row, the second number denotes column
# The first feature for the patient in row 1
brca_clin_df[1,1]

```


```{r}
# The function`c()` combines values into a list
# The colon operator `:` creates a sequence of numbers
c(1:3,8,15)

```

Let's look at the last three features for a group of patients.

```{r}
brca_clin_df[c(1:3,8,15),25:27]

```


### Practice

Change the index numbers to explore the data frame. 


#### Practice 1: Look at a single element

```{r}
# Look at a single element
# The first number denotes row, the second number denotes column
brca_clin_df[100,15]

```



#### Practice 2: Look at the data for a single patient.

```{r}
# Look at a single row or patient
# `t()` transposes the data 
t(brca_clin_df[1,])

```



#### Practice 3: Look at ranges of values using the the combine function and the colon operator.

```{r}
# Examine a set of features for a group of patients
brca_clin_df[c(20,30,40),1:5]

```



#### Practice 4: More meaningful output

**Readability**: We can combine functions with text to get more informative output.

Remember that the `dim()` function provides two numbers:

- The number of rows
- The number of columns

```{r}
# Assign the dimensions of the data frame to an object.

dim_df <- dim(brca_clin_df)  

# Since `dim_df` has two values, we can choose each separately with indexing.
# The function paste() allows us to print text and numbers together.

print(paste("The clinical data frame has",dim_df[1],
              "rows and",dim_df[2],"columns."))

```

#### Practice 4: Writing files

What are all of the features in the data frame?  We can collect the column names.

```{r}
# Many function names in R are intuitive.
# colnames gives the names of the columns.

colnames(brca_clin_df)

```

Notice there are 27 column names for the 27 columns.


```{r}
# We can save the feature names to a file for reference!
# `write.csv()` saves the output of our command to a csv file 
# in your working directory.

write.csv(colnames(brca_clin_df),"features.csv")

```

Check your directory to see if the file is there.

---

## Examining features in the clinical data frame

Viewing the data frame gave us an idea of the types of information included, but how do we summarize the values for all 1,082 patients for a given feature, such as `gender`, `vital_status`, or `estrogen_receptor_status`? 

### Gender

It looks like all of the patients are "FEMALE", but is this true?

The `table()` function is VERY useful. It summarizes the data for us.

```{r}
# The $ operator can be used to select a column with its name
# We can get the column for gender information with `$gender`.
table(brca_clin_df$gender)

```

There are a dozen male patients in this cohort. 


### Age at Diagnosis

The function `hist()` creates a histogram of numerical data. 
Let's see what values are in the feature `age_at_diagnosis`.

```{r}
# The $ operator can be used to select a column with its name.
hist(brca_clin_df$age_at_diagnosis)

```


As with many measurements in nature, the variable `age_at_diagnosis` follows a normal distribution (i.e. it is bell curve).

In our plot, we should label our axes and add a title!  

```{r}
# The $ operator can be used to select a column with its name.
# `xlab` and `main` are arguments to `hist()`.
hist(brca_clin_df$age_at_diagnosis,
     xlab="Age at Diagnosis",      #label the x-axis
     main="Distribution of Ages")  #Add a title

```


Remember, we can always use the `help()` function to learn more about what a function does and what arguments it takes.

```{r,eval=FALSE}
# Get the help page for the function `hist()`
help(hist)

```



### Estrogen Receptor status

Breast cancer cells are tested to see if they have estrogen receptors. When the hormone estrogen attaches to its receptors, it stimulates the cancer to grow. Cancers are called **estrogen receptor-positive (ER+)** or **estrogen receptor-negative (ER-)** based on whether or not they have these receptors. 

Knowing the estrogen receptor status is important in deciding treatment options.

Many breast cancer tumors are estrogen receptor (ER) positive.  How many ER+ and ER- tumors are in the TCGA breast cancer cohort?  

In the previous section, we saw there is a feature called `estrogen_receptor_status`.

```{r}
# The $ operator can be used to select a column with its name.
table(brca_clin_df$estrogen_receptor_status)

```

Some samples do not have information on receptor status, either "[Not Evaluated]" or "Indeterminate".  

It is usually the case with real-life data that not all records (here, patients) have values for all features.

So when working with complex data sets like this, we have to be aware that there may be missing values. 



### Progesterone Receptor status

Like estrogen, progesterone is a hormone that stimulates breast cancer cells to grow.

Let's look at the status of the progesterone receptor negative (PR).
There's a data frame column called `progesterone_receptor_status`.

```{r}
# The $ operator can be used to select a column with its name.
table(brca_clin_df$progesterone_receptor_status)

```


According to these summaries, there are 796 ER+ samples and 689 PR+ samples. 

How many samples have positive status for both estrogen and progesterone receptors?  In other words, how many samples have both ER+ and PR+?

We want to find the patients (rows in `brca_clin_df`) that have

brca_clin_df$estrogen_receptor_status == "Positive"

**AND**

brca_clin_df$progesterone_receptor_status == "Positive".


```{r}
# The operator == finds instances that contain the value "Positive"
# The & operator ensures we apply both conditions.
# Save the results to a variable called `ERpos_PR_pos` 
ERpos_PRpos <- brca_clin_df[(brca_clin_df$estrogen_receptor_status == "Positive" 
               &
               brca_clin_df$progesterone_receptor_status == "Positive"),  ]

# How many patients (rows) have both ER+ and PR+?
nrow(ERpos_PRpos)

```


The overlap between the 796 ER+ tumors and 689 PR+ tumors is 672. 

*Many breast cancer tumors are both ER+ and PR+.*

We can make this output more readable for our report.  Remember, we already saved this data frame result to the object `ERpos_PRpos`.

```{r}
# The function paste() allows us to print text and numbers together.

print(paste("The number of samples with both ER+ and PR+ is", nrow(ERpos_PRpos)))

```


### HER2 status

HER2 is the gene name for the Receptor tyrosine-protein kinase erbB-2; it helps control ceel growth.

There is a data feature called "her2_receptor_status."  You can see this by using the `colnames()` function as above.

```{r}
# The $ operator can be used to select a column with its name.
table(brca_clin_df$her2_receptor_status)

```

In addition to the missing values we found for ER and PR ([Not Available] and Indeterminate), HER2 status can also be "Equivocal" and "Indeterminate."

---

## Triple Negative Breast Cancer

Triple negative breast cancer samples are 

 + estrogen receptor negative (ER-)
 + progesterone receptor negative (PR-)
 + HER2 negative (HER2-).
 
Triple negative breast cancer is agressive and difficult to treat.

How many samples have *"Negative"* status in all three cases?

The pieces of information we need are in the columns called

 + estrogen_receptor_status
 + progesterone_receptor_status
 + her2_receptor_status


Let's make a data frame `receptor_status` that contains the sample names (`bcr_patient_barcode`) and the three receptor status features. We can do this with indexing or by pulling columns out of the data frame by column name.

```{r}
# The $ operator can be used to select a column with its name.
sample_id <- brca_clin_df$bcr_patient_barcode
er_status <- brca_clin_df$estrogen_receptor_status
pr_status <- brca_clin_df$progesterone_receptor_status
her2_status <- brca_clin_df$her2_receptor_status

# `cbind()` takes each vector as a column in a new data frame
receptor_status<- cbind(sample_id,er_status,pr_status,her2_status)

```


The data frame `receptor_status` contains a subset of the full data frame `brca_clin_df`.

```{r}
# We can perform operations on the new dataframe
dim(receptor_status)
head(receptor_status)

```

From `receptor_status`, we'll create another data frame `tnbc` (for triple negative breast cancer) that contains only those samples that have *"Negative"* status for all three receptors.

```{r}

tnbc <- receptor_status[(receptor_status[,2]=="Negative"&
                        receptor_status[,3]=="Negative"&
                        receptor_status[,4]=="Negative"),]

# Let's check out the new dataframe tnbc
head(tnbc)

nrow(tnbc)

```

So overall, 114 samples out of 1082 total samples correspond to Triple Negative Breast Cancer, the most aggressive type of primary breast cancer.   


### Histology

Create a table for "histological_type". 

```{r}

table(brca_clin_df$histological_type)

```

Most of the breast cancer samples are either "Infiltrating Ductal Carcinoma" (774 samples) or "Infiltrating Lobular Carcinoma" (201 samples), the two most common types of breast cancer.

We can look for a specific instance of an histological type by indexing on type name. 

```{r}
# Save this number to a new object
N_ductal <- table(brca_clin_df$histological_type)["Infiltrating Ductal Carcinoma"]

print(N_ductal)

```

## mini-DREAM Challenge 1

Stage II cancer is growing, but it is still contained in the breast or the growth has extended only to nearby lymph nodes. 

How many patients have stage II cancer?

```{r}
# First find out what the values are for `pathologic_stage`
table(brca_clin_df$pathologic_stage)
```

X denotes indeterminate stage.

```{r,eval}

# Uncomment the code, replace xxxxxx with the correct feature name and yyyyyy with the data fram name for stage 2

# N_stage2 <- table(brca_clin_df$xxxxxx)["YY"]

# print(N_stage2)

```



## mini-DREAM Challenge 2

What is the `tumor_status` across patients?

```{r}
table(brca_clin_df$tumor_status)
```

How many samples are "WITH TUMOR"? 

```{r}
# uncomment the code and assign the correct value to `N_tumor`
# N_tumor <- 

```

How many patients with `tumor_status` = "WITH TUMOR" also have `pathologic.stage` = "II"? 

First, create a data frame with the samples that meet these criteria.

```{r}
# Uncomment the next three lines of code and replace YY and XXXXX with the correct values

# tumor_II<-brca_clin_df[(brca_clin_df$pathologic_stage=="YY"
#                 &
#                  brca_clin_df$tumor_status=="XXXXX"),]

```

Then, calculate the size of this cohort.

```{r}
# Choose a function ffff to determine how many rows (patients) meet this condition

# ffff(tumor_II)
```

Finally, assign your result.

```{r}
# Uncomment and assign the value you obtain

# N_tumor_II <-

```

You will have found that about half of the tumors in this cohort are stage II.  

Stage II breast cancer is usually treated by surgical removal followed by radiation. 

## Wrapping up

**Great work!!**.  

We have learned a lot about breast cancer by examining data systematically collected from patients. It is great that these patients gave their consent to be part of TCGA. The data is **anonymized** so that it cannot be traced back to any particular person. TCGA data is publicly available and used by researchers worldwide to improve our ability to diagnose, treat, and prevent cancer.  


In future sessions, we will be introduced to the **gene expression data for these patients**. 


