---
title: "Breast Cancer Patient Data"
author: "Diana Murray"
date: "June 26, 2022"
output:
  html_document:
    toc: yes
    toc_float: yes
    theme: spacelab
    code_folding: show
  pdf_document:
    toc: yes
---

## About this activity

In this activity, we will put our new skills in R to use with a large real-life dataset!

You will load and examine an [R data frame](https://dream-high.shinyapps.io/Data_Frames_Tutorial/) that contains clinical information from over 1,000 breast cancer patients from The Cancer Genome Atlas (TCGA). 

While [R data frame](https://dream-high.shinyapps.io/Data_Frames_Tutorial/) looks a bit ugly here, in our final document, there will be a nice web link to `R data frame`.

[The Cancer Genome Atlas or TCGA](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga) characterized over 20,000 cancer samples spanning 33 cancer types with genomics. 

In 2012 and 2018, the TCGA Network reported findings from the analysis of tumor samples from hundreds of breast cancer patients. Throughout this course, we will examine some of the different data types that were used and the computational analyses that were performed.  

---

## Preliminaries

#### The knitr R package 

**knitr()** is the R package that generates the report from R Markdown. We can create reports as Word doc, PDF, and HTML files. 

An R package bundles together code, data, documentation, and tests, and is easy to download and share with others.

In RStudio, code chucks appear in grey (by default) and begin and end with triple back ticks. 
We run code chunks by **clicking the green hour in the upper right corner of the chunk**.

This website will help you understand what's going on with [R code chunks](https://rmarkdown.rstudio.com/lesson-3.html). 


```{r setup, include=FALSE}  

# Lines beginning with the hash tag symbol (like this one!) allow us to include 
# comments in our code. This is a VERY important practice because it 
# helps us and others to understand what we are doing.
# Curly braces provide tell us what will happen when we knit our document. 
# In this case include=FALSE prevents code and results from appearing in the finished file.

library(knitr)
knitr::opts_chunk$set(echo = TRUE)

```

#### The data directory

We will create an object that holds the name of the directory where the TCGA data resides. 
This is good R coding practice because we can apply all we do below to a data set 
in a different directory by changing a single variable (`data_dir`).

```{r}

data_dir <- "/home/data"    # The data we are using is located in this directory 
                             # on our cloud computing server.

# Look in your Environment tab!

```


---

## Loading the data

**ASIDE**: Notice that the text portions of this document contain many symbols.  Sections have different numbers of hash tags (#) and variables and inline code have back ticks.  You can get a [*Markdown Cheat Sheet*](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf) to make your own formatting changes.


Throughout the course, we will be loading data from R files with the extension `.RData`. 
An example is the file `brca_clin.RData` which we examine below.

`.RData` files are binary files (i.e., not human readable) that hold R objects, such as [vectors, lists, matrices, and data frames](https://dream-high.org/tutorials). 

The R data frame object in the `brca_clin.RData` file is the result of previous R computations that put raw data downloaded from the TCGA project into **tidy** form. 

Tidy data sets have consistent structure and are easy to manipulate, visualize, and model.
They are "clean" and ready for us to use! 

We could spend the entire course learning how to put raw data in tidy form, but our goal is to understand and analyze the data, which mostly occurs *after* tidying the data.


```{r}

# With the function `load()` we read in the file data file "brca_clin.RData".
# The file.path() function tells `load()` where our data resides.
# The argument "verbose = TRUE" makes `load()` tell us what R objects are being loaded.
# The objects will also appear in our "Environment" tab.

load(file.path(data_dir, "brca_clin.RData"),verbose=TRUE)   

```

The object `brca_clin_df` was named to be be descriptive.

* "brca" stands for "TCGA breast cancer"
* "clin" is short for "clinical data"
* "df" stands for "data frame"

---

## Viewing the data

The **Environment** tab gives us basic information on the objects that were loaded.

There are 1082 observations (patients) with 27 variables. 

We can use some simple R functions to examine the data in more detail.

* `dim()` tells us the dimensions (# row, # columns) of the object.
* `head()` returns the first several rows of the objects.
* **indexing** allows us to look at rows and columns we choose.


```{r}

# How many rows and columns are in our object?
dim(brca_clin_df)

```

Is the output what we expect? Check your **Environment** again. 


```{r}

# Let's take a first look at the data in our object
# `head` gives us the first six rows and ALL of the columns

head(brca_clin_df)

```

Since our data frame has many columns, we can scroll through the columns by clicking on the black arrow in the top right corner.

In data science, columns are also referred to as [features or feature variables](https://www.datarobot.com/wiki/feature/). 

Notice the column names. Let's define a few:
 + `bcr_patient_barcode`  provides anonymous identifiers for the patients.
 + `estrogen_receptor_status`  designates whether the tumor is ER+ or ER-
 + `progesterone_receptor_status` designates whether the tumor is PR+ or PR-
 + `her2_receptor_status`  designates whether the tumor is HER2+ or HER2- 

We will be using these features in future hands-on coding sessions. In this session, we will explore the information they contain.

**ASIDE**: We can combine functions with text to get more informative output:

```{r}

dim_df <- dim(brca_clin_df)  # We should get the same result as before.

# Since `dim_df` has two values, we can choose each separately with indexing.

print(paste("The clinical data frame has",dim_df[1],
              "rows and",dim_df[2],"columns."))

# The function paste() allows us to print text and numbers together.

```

This is more comprehensible for our report!


Since this is a relatively small data frame (1082 rows is small compared to the gene expression data frames we will consider!), we can use the `View()` function to examine it in detail. 

`View()` will open the entire object in a new tab.

```{r include=FALSE}

# Uncomment the next line by removing the `#` symbol to `View()` the data
# View(brca_clin_df)  

```

Notice that the data frame opens in a new window with the title `brca_clin_df`.

Wow! That's a lot of information.  But it's only a fraction of the information available from TCGA.

What are the features in the data frame?  We can collect the column names.

```{r}

# Many function names in R are intuitive.
# colnames gives the names of the columns.

colnames(brca_clin_df)

```


```{r}

# We can save the feature names to a file for reference!
# `write.csv` saves the output of our command to a csv file 
# in your working directory.

# Uncomment the next line to run the code.

# write.csv(colnames(brca_clin_df),"features.csv")

```

Check your directory to see if the file is there.

---

## Examining the Features

Examining the data frame in tabular form with `View()` gives us an idea of the types of information included, but it's hard to take in, by eye, the values for all 1,082 patients for any given feature, such as `gender`, `vital_status`, or `estrogen_receptor_status`. 

### Gender

For example, it looks like all of the patients are "FEMALE", but is this true?

The `table()` function is VERY useful. It summarizes the data for us.

```{r}

# We can get a column by designating its column index.

table(brca_clin_df[,2])

# Or we can use the column name, for example, with `$gender`.

table(brca_clin_df$gender)

```

There are a dozen male patients in this cohort. 

A Patient Advocate, Bob Ritter, who works with the Cancer System Biology Consortium (CSBC) and the Physical Sciences in Oncology (PS-ON) Program talks about his experience with breast cancer.  

You can read more about Bob's amazing and courageous story here [Bob Riter: The Male Breast Cancer Coalition](https://malebreastcancercoalition.org/Survivor%20Stories/bob-riter/).


### Age at Diagnosis

The function `hist` will give us a histogram of numerical data. 
Let's see what values are in the feature `age_at_diagnosis`.

```{r}

hist(brca_clin_df$age_at_diagnosis)

```


As many measurements in nature, the variable `age_at_diagnosis` follows a normal distribution (i.e. it is bell curve).

This is nice, but we should label our axes!  
We can do this by adding *argument* to the function `hist()`.


```{r}

hist(brca_clin_df$age_at_diagnosis,
     xlab="Age at Diagnosis",
     main="Distribution of Ages")

```

Remember, we can always use the `help()` functions to learn more about what a function does and what arguments it takes.

```{r}

help(hist)

```



### Estrogen Receptor status

Many breast cancer tumors are estrogen receptor (ER) positive.  How many ER+ and ER- tumors are in our dataset?  

We saw above that there's a data frame feature called `estrogen_receptor_status`.

```{r}

table(brca_clin_df$estrogen_receptor_status)

```

Some samples do not have information on receptor status (either "[Not Evaluated]" or "Indeterminate").  

It is usually the case with real-life data that not all records (here, patients) have values for all features in the data frame.

So when working with complex data sets like this, we have to be aware that there may be missing values. We will continually be on the lookout for **missing values** in our activities!


### Progesterone Receptor status

Let's look at the status of the progesterone receptor negative (PR).
There's a data frame column called `progesterone_receptor_status`.

```{r}

table(brca_clin_df$progesterone_receptor_status)

```


According to these summaries, there are 796 ER+ samples and 689 PR+ samples. 

How many samples have positive status for both estrogen and progesterone receptors?  
In other words, how many samples have both ER+ and PR+?

We want to find the patients (rows in `brca_clin_df`) with 

brca_clin_df$estrogen_receptor_status == "Positive"

AND

brca_clin_df$progesterone_receptor_status == "Positive".

We use `==` to find the cases that have the value "Positive", and

`&` takes only those samples that meet **both** criteria.

The function `nrow()` will give us the number of rows. 

Let's put the pieces together:

```{r}

# We apply the conditions to the rows.
# We save the results to a variable called `ERpos_PR_pos` 
# which allow us to use the data later.

ERpos_PRpos <- brca_clin_df[(brca_clin_df$estrogen_receptor_status == "Positive" 
               &
               brca_clin_df$progesterone_receptor_status == "Positive"),  ]

nrow(ERpos_PRpos)

```

We could also do this in one step:

```{r}

nrow(brca_clin_df[(brca_clin_df$estrogen_receptor_status == "Positive" &
                   brca_clin_df$progesterone_receptor_status == "Positive"),  ])

```

Overall, the overlap between the 796 ER+ tumors and 689 PR+ tumors is 672. 


We can make this output more readable for our report!  Remember, we already saved this data frame result to the object `ERpos_PRpos`.

```{r}


print(paste("The number of samples with both ER+ and PR+ is ", nrow(ERpos_PRpos)))

# Again, the function paste() allows us to print text and numbers together.

```

Many breast cancer tumors are *both* ER+ and PR+.


### HER2 status

HER2 is stands for the Receptor tyrosine-protein kinase erbB-2.

There's a data frame column called "her2_receptor_status".

```{r}

table(brca_clin_df$her2_receptor_status)

```

In addition to the missing values we found for ER and PR ([Not Available] and Indeterminate), HER2 status can also be "Equivocal" and "Indeterminate." 

But still, we are interested mainly in "Positive" and "Negative."

---

## Triple Negative Breast Cancer

Triple negative breast cancer samples are 

 + estrogen receptor negative (ER-)
 + progesterone receptor negative (PR-)
 + HER2 negative (HER2-).

How many samples have "Negative" status in all three cases?

As we see from the data frame itself and the three code chunks above, the pieces of information we need are in the columns called

 + estrogen_receptor_status
 + progesterone_receptor_status
 + her2_receptor_status


Let's make a data frame `receptor_status` that contains the sample names (`bcr_patient_barcode`) and the three receptor status features.

```{r}
sample_id <- brca_clin_df$bcr_patient_barcode
er_status <- brca_clin_df$estrogen_receptor_status
pr_status <- brca_clin_df$progesterone_receptor_status
her2_status <- brca_clin_df$her2_receptor_status

receptor_status<- cbind(sample_id,er_status,pr_status,her2_status)

# Having the results in a data frame object allows us to do operations on it.

dim(receptor_status)
head(receptor_status)
```


Let's organize the output to be a bit more readable with *newlines* `\n`.


```{r}

sample_id <- brca_clin_df$bcr_patient_barcode
er_status <- brca_clin_df$estrogen_receptor_status
pr_status <- brca_clin_df$progesterone_receptor_status
her2_status <- brca_clin_df$her2_receptor_status

receptor_status<- cbind(sample_id,er_status,pr_status,her2_status)

# Having the results in a data frame object allows us to do operations on it.

print("We created a new data frame `receptor_status` that contains onle the patient ids and the features for the three receptors ER, PR, and HER2.")

cat("\n")  # This function adds a blank line in the output.

print(paste("The object `receptor_status` has ",dim(receptor_status)[1], 
            "rows and ", dim(receptor_status)[1]," columns."))

cat("\n")  # This function adds a blank line in the output.

print("The first few rows of `receptor_status` are:")

head(receptor_status)
```


The data frame `receptor_status` contains a subset of the full data frame `brca_clin_df`.

From `receptor_status`, we'll create another dat aframe `tnbc` (for triple negative breast cancer) that contains only those samples that have "Negative" status for all three receptors.

```{r}

tnbc <- receptor_status[(receptor_status[,2]=="Negative"&
                        receptor_status[,3]=="Negative"&
                        receptor_status[,4]=="Negative"),]

head(tnbc)

nrow(tnbc)

```

So overall, 114 samples out of 1082 total samples correspond to Triple Negative Breast Cancer, the most aggressive type of primary breast cancer. And we know which samples those are because we recorded this information in the object `tnbc`. 


Since we know `sample_id` for the samples with Triple Negative Breast Cancer, we can add additional information about them to our object `tnbc`.

First, we'll use the function`match()` the `sample_id` column in `tnbc` with the `bcr_patient_barcode` column in `brca_clin_df`.

```{r}

a <- match(tnbc[,1],brca_clin_df[,1])
head(a)

```

These numbers are the rows that correspond to the samples in `tnbc`. 

Let's add the `gender` and `vital_status` columns (columns 2 and 7) from `brca_clin_df` for these samples to `tnbc`.


```{r}

tnbc <- cbind(tnbc, brca_clin_df[a,c(2,7)])   

# `a` is a vector of the row numbers we want.
# c(2,7) collects the information from columns 2 and 7


head(tnbc)

```


You can use `View()` to see the information for all 114 samples in `tnbc`.


```{r include=FALSE}

# Uncomment the next line to `View()` the data
# View(tnbc)  

```

---

## Practice

Feel free to modify the code chunks above or to create new code chunks!

#### Practice 1

Examine other features in `brca_clin_df` using the `table()` function and either indexing or specifying the column name.

For example, both commands in the next code chunk provide tables of the information in column 16 which corresponds to "histological_type". 


```{r}

table(brca_clin_df[,15])

table(brca_clin_df$histological_type)

```

We see that most of the breast cancer samples are either "Infiltrating Ductal Carcinoma" or "Infiltrating Lobular Carcinoma", the two most common types of breast cancer.

How do we get **just** the number of samples with `Infiltrating Ductal Carcinoma`?

```{r}

# Look at the table to see that `Infiltrating Ductal Carcinoma` 
# is the third item in the table

# View(table(brca_clin_df$histological_type))

```

```{r}

# Save this number to a new object

N_ductal <- table(brca_clin_df$histological_type)[3]

print(N_ductal)

```

#### mini-DREAM Challenge submission part 1

How many samples are of histological_type "Infiltrating Lobular Carcinoma"?

Save your result to an object called N_lobular.

```{r}

# Uncomment the code and replace N with the correct index for "Infiltrating Lobular Carcinoma"

# N_lobular <- table(brca_clin_df$histological_type)[N]


```


#### mini-DREAM Challenge submission part 2

How many samples are "TUMOR FREE"? 


```{r}

# First find out which feature provides information on "TUMOR FREE"
# Uncomment

# View(brca_clin_df)


```


Uncomment one of the lines in the code chunk to get the summary information that gives the number of "TUMOR FREE" samples

```{r}

#my_data_frame <- table((brca_clin_df$xxxxx_yyyyyy))


#my_data_frame <- table((brca_clin_df[,NN]))

```


Save the number for "TUMOR FREE" samples to the object `N_tumor_free` by replacing MMM.

```{r}
# Uncomment the code to run it

# N_tumor_free <- MMM


```


Submit "N_lobular" and "N_tumor_free" to the leaderboard.

```{r}
# Load function to submitting answers to the leader board
scripts_dir <- "/home/shared/R"
source(file.path(scripts_dir, "submission_helpers.R"))

# Log into Synapse
synLoginSecure()  # You might be prompted for your username and password

# Submit answers
submission <- submit_module_answers(module = 1)
```

You can now return to the **mini-DREAM** Challenge site on Synapse. **Make sure to save your work before you leave!**

**Great work!!**.  We have learned a lot about breast cancer by examining data systematically collected from patients. It is great that these patients gave their consent to be part of TCGA. The data is **anonymized** so that it cannot be traced back to any particular person. TCGA data is publicly available and used by researchers worldwide to improve our ability to diagnose, treat, and prevent cancer.  

In future sessions, we will be introduced to the gene expression data for these patients. 


#### BONUS Practice 1

Do the same with other features in `brca_clin_df`.

```{r eval = FALSE}

# Replace NN with the column number you're interested in.
# Remember to uncomment the code lines!!

# table(brca_clin_df[,NN])

# Replace column_name with the name of the column in `brca_clin_df`

# table(brca_clin_df$column_name)

```



#### BONUS: Practice 2

Create a data frame for the ER+ and PR+ patient samples.  In the code chucks for triple negative breast cancer, we created the variables `sample_id`, `er_status` and `pr_status`, and used the function `cbind()` to create a data frame object called `receptor_status`. Take a look at that object again:

```{r eval = FALSE}

dim(receptor_status)

head(receptor_status)

```


Create a new object `hormone_receptor_positive` that contains only the samples where er_status and pr_status are "Positive".

```{r eval = FALSE}

# Replace "Feature_value" with the value you want.

hormone_receptor_positive<- receptor_status[(receptor_status[,2]=="Feature_value"&
                                             receptor_status[,3]=="Feature_value"),]
                        
head(hormone_receptor_positive)

# Uncomment the `View()` command to see the entire object
# View(hormone_receptor_positive)

```

Notice the values for her2_status in `hormone_receptor_positive`. 


Finally, try adding additional features to your object `hormone_receptor_positive` following what we did for the object `tnbc` we created.


```{r eval = FALSE}



```



### BONUS: Practice 3

There are hundreds of lymph nodes throughout the body. Lymph nodes are part of the lymphatic system which, in turn, is part of the immune system. Lymph nodes are little bean-shaped structures that contain immune cells that help fight infection. They’re connected by a system of lymph vessels that carry fluid throughout your body. 

Cancer spreads when cancer cells break away from the primary tumor. In breast cancer, these cells are most likely to reach the lymph nodes closest to the affected breast.

Breast cancer has four stages. When lymph nodes are involved, it’s at least stage 2. Metastatic breast cancer is stage 4.

There are some samples where `lymph_node_examined_count` is "[Not Available]".
Let's remove those.

```{r}

LN_counted<-brca_clin_df[((brca_clin_df$lymph_node_examined_count)!="[Not Available]"),]

range(as.numeric(LN_counted$lymph_node_examined_count))


```


What percentage of samples have information on lymph node count?

```{r}

nrow(LN_counted)/nrow(brca_clin_df) * 100

```

89% of the samples have associated lymph node data.

Let's get a sense of how many lymph nodes were examined in each sample:

```{r}

hist(as.numeric(LN_counted$lymph_node_examined_count),
     xlab="Number of lymph nodes examined",
     main="Lymph nodes in 88% of patient samples were tested",
     breaks=20)

```

If you've come this far, your knowledge of R is impressive! **Congratulations!!**



